File: backend/src/cart/core/CartItem.ts
/**
 * Represents a CartItem domain entity.
 */
export class CartItem {
  constructor(
    public readonly id: string,
    public readonly cartId: string,
    public readonly productVariantId: string,
    public readonly quantity: number,
    public readonly unitPriceSnapshot: number, // in cents
    public readonly createdAt: string,
    public readonly updatedAt: string,
  ) {}

  public static create(params: {
    id: string;
    cartId: string;
    productVariantId: string;
    quantity: number;
    unitPriceSnapshot: number;
    createdAt?: string;
    updatedAt?: string;
  }): CartItem {
    const now = new Date().toISOString();
    return new CartItem(
      params.id,
      params.cartId,
      params.productVariantId,
      params.quantity,
      params.unitPriceSnapshot,
      params.createdAt ?? now,
      params.updatedAt ?? now,
    );
  }
}

File: mobile/lib/features/cart/domain/entities/cart_item.dart
/// CartItem domain entity on mobile.
class CartItem {
  final String id;
  final String cartId;
  final String productVariantId;
  final int quantity;
  final int unitPriceSnapshot; // in cents
  final String productName;
  final String size;
  final String color;
  final String sku;
  final String imageUrl;

  const CartItem({
    required this.id,
    required this.cartId,
    required this.productVariantId,
    required this.quantity,
    required this.unitPriceSnapshot,
    required this.productName,
    required this.size,
    required this.color,
    required this.sku,
    required this.imageUrl,
  });

  /// Calculates the total cost for this item line.
  int get subtotal => quantity * unitPriceSnapshot;
}
